src/dataflow/src/source/mod.rs

```diff
@@ -730,11 +734,52 @@ pub struct PartitionMetrics {
     closed_ts: DeleteOnDropGauge<'static, AtomicU64>,
     /// Total number of messages that have been received by the source and timestamped
     messages_ingested: DeleteOnDropCounter<'static, AtomicI64>,
+    logger: Option<Logger>,
+    source_name: String,
+    source_id: String,
+    partition_id: String,
+    last_received_offset: i64,
+    last_ingested_offset: i64,
 }
 
 impl PartitionMetrics {
+    /// Record the latest offset received high-water mark
+    pub fn record_offset_received(&mut self, offset: i64) {
+        self.offset_received.set(offset);
+        if let Some(logger) = self.logger.as_mut() {
+            logger.log(MaterializedEvent::SourceInfo(
+                self.source_name.clone(),
+                self.source_id.clone(),
+                self.partition_id.clone(),
+                offset - self.last_received_offset,
+                0,
+            ));
+        }
+        self.last_received_offset = offset;
+    }
+
+    /// Record the latest offset ingested high-water mark
+    pub fn record_offset_ingested(&mut self, offset: i64) {
+        self.offset_ingested.set(offset);
+        if let Some(logger) = self.logger.as_mut() {
+            logger.log(MaterializedEvent::SourceInfo(
+                self.source_name.clone(),
+                self.source_id.clone(),
+                self.partition_id.clone(),
```

Comment:
this is going to cause three allocations on every kafka batch, which is hot enough that we refactored to the current PartitionMetrics struct to relieve pressure on the allocator.

I'm not sure if it's possible for us to turn this into `logger.log(Arc<SourceInfo>)` or something similar? Possibly otherwise we should use [`smol_str`](https://github.com/rust-analyzer/smol_str) here, if that's doable inside of timely.